src/workclaimer.js

```diff
@@ -0,0 +1,298 @@
+let assert      = require('assert');
+let _           = require('lodash');
+let Promise     = require('promise');
+let events      = require('events');
+let taskcluster = require('taskcluster-client');
+
+/**
+ * HintPoller polls for hints for pending tasks.
+ *
+ * The azure queues don't know if a task is pending they just store hints of
+ * pending tasks. To be understood this way:
+ *  A) If a task is pending, there is a hint of the task in an azure queue,
+ *  B) If there is an hint in an azure queue, it may or may not be pending.
+ *
+ * It's an if, but not an only-if (think over-approximation).
+ */
+class HintPoller {
+  constructor(parent, provisionerId, workerType) {
+    this.parent = parent;
+    this.provisionerId = provisionerId;
+    this.workerType = workerType;
+    this.requests = [];
+    this.started = false;
+  }
+
+  requestClaim(count, aborted) {
+    // Make a request for count tasks
+    let request = null;
+    let result = new Promise((resolve, reject) => {
+      request = {resolve, reject, count};
+    });
+    this.requests.push(request);
+
+    aborted.then(() => {
+      // Remove request from requests, and resolve request empty array
+      _.remove(this.requests, request);
+      request.resolve([]);
+    });
+
+    this.start();
+
+    return result;
+  }
+
+  start() {
+    if (!this.started) {
+      this.started = true;
+      this.poll().catch(err => {
+        this.started = false;
+        // Resolve everything as failed
+        this.destroy();
+        let requests = this.requests;
+        this.requests = [];
+        requests.map(r => r.reject(err));
+      }).catch(err => {
+        process.nextTick(() => this.parent.emit('error', err));
+      });
+    }
+  }
+
+  async poll() {
+    // Get poll functions for pending queues (ordered by priority)
+    let polls = await this.parent._queueService.pendingQueues(
+      this.provisionerId, this.workerType,
+    );
+    // While we have requests for hints
+    while (_.sumBy(this.requests, 'count') > 0) {
+      let claimed = 0; // Count hints claimed
+
+      // In-order of priority, we poll hints from queues
+      for (let poll of polls) {
```

Comment:
> we agreed that there should be some sort of priority system better than all of the top priority then all of the next priority... stuff. I think that's what's happening here.

Next time you need to get that in writing :) hehe -- I recall the discussion I don't recall agreeing that we need nice-levels as oppose to strict priority levels. I had a discussion with catlee where he told me the current bb setup is strict priority.

Regardless, I maintain that if we want to shuffle the queues randomly every once in a while we can do so... But I want to see it being a requested feature.

TL;DR: We can do this later if we really want to. I don't think we want nice levels. But I'm not blocking in this on that discussion.